frappe/public/js/frappe/form/multi_select_dialog.js

```diff
@@ -0,0 +1,199 @@
+
+		this.dialog = new frappe.ui.Dialog({
+			title: __("Select {0}", [(this.doctype=='[Select]') ? __("value") : __(doctype_plural)]),
+			fields: fields,
+			primary_action_label: __("Get Items"),
+			primary_action: function() {
+				me.action(me.selections, me.args);
+			}
+		});
+
+		this.$parent = $(this.dialog.body);
+		this.$wrapper = this.dialog.fields_dict.results_area.$wrapper.append(`<div class="results"
+			style="border: 1px solid #d1d8dd; border-radius: 3px; height: 300px; overflow: auto;"></div>`);
+		this.$results = this.$wrapper.find('.results');
+		this.$make_new_btn = this.dialog.fields_dict.make_new.$wrapper;
+		this.$placeholder = $(`<div class="multiselect-empty-state"> <span class="text-center" style="margin-top: -40px;">
+			<i class="fa fa-2x fa-tags text-extra-muted"></i> <p class="text-extra-muted">No ${this.doctype} found</p>
+			<button class="btn btn-default btn-xs text-muted" data-fieldtype="Button" data-fieldname="make_new"
+			placeholder="" value="">Make a new ${this.doctype}</button></span> </div>`);
+
+		this.selections = [];
+		this.args = {};
+
+		this.bind_events();
+		this.get_results();
+		this.dialog.show();
+	},
+
+	bind_events: function() {
+		let me = this;
+
+		this.$results.on('change', '.list-item-container :checkbox', function (e) {
+			var $item = $(this).closest('.list-item-container');
+			var filename = $item.attr('data-filename');
+			var $target = $(e.target);
+
+			var checked = $target.is(':checked');
+			if (checked){
+				me.selections.push(filename);
+			} else {
+				var index = me.selections.indexOf(filename);
+				if (index > -1) me.selections.splice(index, 1);
+			}
+		});
+		this.$results.on('click', '.list-item-container', function (e) {
+			if (!$(e.target).is(':checkbox') && !$(e.target).is('a')) {
+				$(this).find(':checkbox').trigger('click');
+			}
+		});
+		this.$results.on('click', '.list-item--head :checkbox', function (e) {
+			if ($(e.target).is(':checked')) {
+				me.$results.find('.list-item-container :checkbox:not(:checked)').trigger('click');
+			} else {
+				me.$results.find('.list-item-container :checkbox(:checked)').trigger('click');
+			}
+		});
+
+		this.$parent.find('.input-with-feedback').on('change', function(e) {
+			me.get_results();
+		});
+		this.$parent.on('click', '.btn[data-fieldname="make_new"]', function(e) {
+			frappe.route_options = {};
+			Object.keys(me.setters).forEach(function(setter) {
+				frappe.route_options[setter] = me.dialog.fields_dict[setter].get_value() || undefined;
+			});
+			frappe.new_doc(me.doctype, true);
+		});
+	},
+
+	make_list_row: function(result={}) {
+		var me = this;
+		// Make a head row by default (if result not passed)
+		let head = Object.keys(result).length === 0;
+
+		let columns = ["name"].concat(Object.keys(this.setters));
+		let $row = $(`<div class="list-item">
+			<div class="list-item__content ellipsis" style="flex: 0 0 10px;">
+				<input type="checkbox"/>
+			</div>
+			${	(() => {
```

Comment:
please make this a generic function - too much nested functionality - this can be used other places too....